tower-http/src/classify.rs

```diff
@@ -0,0 +1,316 @@
+    /// errors. A retry policy might allows retrying some errors and not others.
+    fn classify_error(self, error: &E) -> Self::FailureClass;
+}
+
+/// Result of doing a classification.
+#[derive(Debug)]
+pub enum ClassifiedResponse<FailureClass, ClassifyEos> {
+    /// The response was able to be classified immediately.
+    Ready(Result<(), FailureClass>),
+    /// We have to wait until the end of a streaming response to classify it.
+    RequiresEos(ClassifyEos),
+}
+
+/// A [`ClassifyEos`] type that can be used in [`ClassifyResponse`] implementations that never have
+/// to classify streaming responses.
+///
+/// `NeverClassifyEos` exists only as type. You cannot construct `NeverClassifyEos` values.
+pub struct NeverClassifyEos<T> {
+    _output_ty: PhantomData<fn() -> T>,
+    _never: Infallible,
+}
+
+impl<T, E> ClassifyEos<E> for NeverClassifyEos<T> {
+    type FailureClass = T;
+
+    fn classify_eos(self, _trailers: Option<&HeaderMap>) -> Result<(), Self::FailureClass> {
+        // `NeverClassifyEos` contains an `Infallible` so it can never be constructed
+        unreachable!()
+    }
+
+    fn classify_error(self, _error: &E) -> Self::FailureClass {
+        // `NeverClassifyEos` contains an `Infallible` so it can never be constructed
+        unreachable!()
+    }
+}
+
+/// The default classifier used for normal HTTP responses.
+///
+/// Responses with a `5xx` status code are failures, all others as successes.
+#[derive(Clone, Debug, Default)]
+pub struct ServerErrorsAsFailures {
+    _priv: (),
+}
+
+impl ServerErrorsAsFailures {
+    /// Create a new [`ServerErrorsAsFailures`].
+    pub fn new() -> Self {
+        Self::default()
+    }
+
+    /// Returns a [`MakeClassifier`] that produces `ServerErrorsAsFailures`.
+    ///
+    /// This is a convenience function that simply calls `SharedClassifier::new`.
+    pub fn make_classifier<E>() -> SharedClassifier<Self> {
+        SharedClassifier::new::<E>(Self::new())
+    }
+}
+
+impl<E> ClassifyResponse<E> for ServerErrorsAsFailures {
+    type FailureClass = StatusCode;
+    type ClassifyEos = NeverClassifyEos<StatusCode>;
+
+    fn classify_response<B>(
+        self,
+        res: &Response<B>,
+    ) -> ClassifiedResponse<Self::FailureClass, Self::ClassifyEos> {
+        if res.status().is_server_error() {
+            ClassifiedResponse::Ready(Err(res.status()))
+        } else {
+            ClassifiedResponse::Ready(Ok(()))
+        }
+    }
+
+    fn classify_error(self, _error: &E) -> Self::FailureClass {
+        StatusCode::INTERNAL_SERVER_ERROR
+    }
+}
+
+/// Response classifier for gRPC responses.
+///
+/// gRPC doesn't use normal HTTP statuses for indicating success or failure but instead a special
+/// header that might appear in a trailer.
+///
+/// Responses where `grpc-status` is 0 are successes, all others are failures.
+#[derive(Debug, Clone, Default)]
+pub struct GrpcErrorsAsFailures {
+    _priv: (),
+}
```

Comment:
afaict this thing's job is just to extract the gRPC status code from the trailers, and handle the cases where its missing or the body errors? it might be worth making this clear in the docs?

hypothetically i can see this getting chained/mapped with another classifier for determining whether or not those status codes are retryable, as part of a retry policy? i think it could be good to add combinators for response classifiers in a future branch so that it's easy to express things like that...